### PR TITLE
SWATCH-803: CRUD apis for swatch-contracts

### PR DESCRIPTION
**Pull request by lindseyburnett**
GET, POST, and DELETE endpoints are working.  PUT will be completed in a separate story.

`./gradlew :swatch-contracts:quarkusDev`

Swagger doc can be found at: http://localhost:8000/q/swagger-ui/

https://petstore.swagger.io/?url=https://raw.githubusercontent.com/RedHatInsights/rhsm-subscriptions/lburnett/swatch-803-contracts-crud-apis/swatch-contracts/src/main/resources/swatch_contracts_openapi.yaml#/ 

# Some curl commands

```bash
curl -X 'POST' \
  'http://localhost:8000/api/swatch-contracts/internal/contracts' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "subscription_number": "subscription456",
  "sku": "BAS_SKU",
  "start_date": "2022-03-10T12:15:50-04:00",
  "org_id": "org123",
  "billing_provider": "aws",
  "billing_account_id": "billingAcct123",
  "product_id": "BASILISK",
  "metrics": [
    {
      "metric_id": "instance-hours",
      "value": 720
    }
  ]
}'
```

```bash
curl -X 'GET' \
  'http://localhost:8000/api/swatch-contracts/internal/contracts' \
  -H 'accept: application/json'
```

```bash
curl -X 'DELETE' \
  'http://localhost:8000/api/swatch-contracts/internal/contracts/2d44ce3b-de01-4d83-a7c8-c0157f3e71d3' \
  -H 'accept: */*'
```

### CODE REVIEW

File: swatch-contracts/src/main/java/com/redhat/swatch/ContractService.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    var newRecord = mapper.dtoToContract(dto);
+    newRecord.setUuid(newUuid);
+    newRecord.setLastUpdated(OffsetDateTime.now());
+    newRecord.setEndDate(null);
+    newRecord.getMetrics().stream()
+        .forEach(
+            metric -> {
+              metric.setContractUuid(newUuid);
+              metric.setContract(newRecord);
+            });
```
[the remark this answers is not in the document]
Comment (mstead) [reply swatch-contracts/src/main/java/com/redhat/swatch/ContractService.java:132]: Found a card that appears to cover this. Please ignore.

File: swatch-contracts/src/main/java/com/redhat/swatch/ContractService.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    var newData = mapper.dtoToContract(dto);
+    newData
+        .getMetrics()
+        .forEach(
+            metric -> {
+              metric.setContractUuid(newData.getUuid());
+              metric.setContract(newData);
+            });
```
[the remark this answers is not in the document]
Comment (mstead) [reply swatch-contracts/src/main/java/com/redhat/swatch/ContractService.java:103]: Found a card that appears to cover this. Please ignore.

**Review by mstead — CHANGES_REQUESTED** [review]
I tried to check existing cards to see if anything I've commented on will be covered elsewhere, so I apologize in advance if I missed them. :)

Some general comments:
* Would it make sense to describe 'local deploy with bonfire' in a README?
* Personally I'd like to see some additional package structure. i.e Can we move the DB related classes to its own package? Could create an additional card to do this, or serve as a discussion point.
* There are some test specific annotations being used which could benefit from some documentation somewhere (mostly around intent so that others can follow suit when needed)
   * `@DisabledIfEnvironmentVariable(named = "CI", matches = "true")`
   * `@Tag("integration")`

**Comment by lindseyburnett** [comment]
> * Would it make sense to describe 'local deploy with bonfire' in a README?
I think our main README.md has a section about this.  Might need updating.

> * Personally I'd like to see some additional package structure. i.e Can we move the DB related classes to its own package? Could create an additional card to do this, or serve as a discussion point.
I added this as a dev sync agenda item.  Seeing things lately in quarkus apps where people are moving away from a bunch of separate packages.

File: swatch-contracts/src/test/java/com/redhat/swatch/ContractServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    ContractMetric expectedMetric2 = new ContractMetric();
+    expectedMetric2.setMetricId("cpu-hours");
+    expectedMetric2.setValue(5);
+    var expected = com.redhat.swatch.Contract.builder().metrics(Set.of(expectedMetric2)).build();
+    var actual = contractService.createContractForLogicalUpdate(dto);
+
+    // new.uuid != old.uuid
+    // new.uuid == new.metrics[].uuid
+    // new.endDate == null
+    // new.startDate == old.endDate
```
[the remark this answers is not in the document]
Comment (kartikshahc) [reply swatch-contracts/src/test/java/com/redhat/swatch/ContractServiceTest.java:166]: Will take care in next card

File: swatch-contracts/src/test/java/com/redhat/swatch/ContractServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    parameters.put("productId", "BASILISK123");
+    List<com.redhat.swatch.openapi.model.Contract> contractList =
+        contractService.getContracts(parameters);
+    verify(contractRepository).getContracts(parameters);
+    assertEquals(1, contractList.size());
+    assertEquals(2, contractList.get(0).getMetrics().size());
+  }
+
+  /*  @Test
+  void testUpdateContract() {
```
[the remark this answers is not in the document]
Comment (kartikshahc) [reply swatch-contracts/src/test/java/com/redhat/swatch/ContractServiceTest.java:131]: :) Will take care in following card

File: swatch-contracts/src/test/java/com/redhat/swatch/BaseUnitTest.java [hunk 5]
```diff
@@ -20,9 +20,13 @@
...
 
-@QuarkusIntegrationTest
-public class ExampleResourceIT extends ExampleResourceTest {
-  // Execute the same tests but in packaged mode.
+@SuppressWarnings("java:S2187") /* Sonar thinks assertion is required */
+class BaseUnitTest {
+  @BeforeEach
+  void init_mocks() {
+    MockitoAnnotations.openMocks(this);
+  }
```
[the remark this answers is not in the document]
Comment (kartikshahc) [reply swatch-contracts/src/test/java/com/redhat/swatch/BaseUnitTest.java:31]: Done

**Pull request merged**